security-csrf/src/main/java/io/micronaut/security/csrf/resolver/HttpHeaderCsrfTokenResolver.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.csrf.resolver;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.security.csrf.CsrfConfiguration;
+import jakarta.inject.Singleton;
+
+import java.util.Optional;
+
+/**
+ * Resolves a CSRF token from a request HTTP Header named {@link CsrfConfiguration#getHeaderName()}.
+ * @author Sergio del Amo
+ * @since 4.11.0
+ */
+@Requires(classes = HttpRequest.class)
+@Requires(property = "micronaut.security.csrf.token-resolvers.http-header.enabled", value = StringUtils.TRUE, defaultValue = StringUtils.TRUE)
+@Singleton
+@Internal
+final class HttpHeaderCsrfTokenResolver implements CsrfTokenResolver<HttpRequest<?>> {
+    private static final int ORDER = -100;
+    private final CsrfConfiguration csrfConfiguration;
+
+    HttpHeaderCsrfTokenResolver(CsrfConfiguration csrfConfiguration) {
+        this.csrfConfiguration = csrfConfiguration;
+    }
+
+    @Override
+    public Optional<String> resolveToken(HttpRequest<?> request) {
+        String csrfToken = request.getHeaders().get(csrfConfiguration.getHeaderName());
```

Comment:
extract `request.getHeaders()` into a local variable to avoid repeated calls

Reply:
resolved via https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/dee17cf0e85bed034667e36c79a6a8c4dfc31253